lla_to_ecef.cpp

```diff
@@ -0,0 +1,390 @@
+ *                  - latitude [degrees]
+ *                  - longitude [degrees]
+ *                  - altitude data [meters]
+ * @param a       reference ellipsoid semi-major axis [meters]
+ * @param b       reference ellipsoid semi-minor axis [meters]
+ * @param e       reference ellipsoid eccentricity [meters]
+ *
+ * @return a PositionVelocityECEF struct with the following:
+ *                  - same as LLA position: time as timespec struct, since Unix epoch
+ *                  - x, y, and z ECEF position fields as calculated, in meters
+ *                  - x, y, and z ECEF velocities are left uninitialized
+ */
+PositionVelocityECEF lla_to_ecef_pos(
+    PositionLLA pos_lla, double a, double b, double e) {
+    double r_curve = get_rad_of_curvature(a, e, pos_lla.latitude);
+
+    double c_lat = std::cos(pos_lla.latitude * DEG_TO_RAD);
+    double s_lat = std::sin(pos_lla.latitude * DEG_TO_RAD);
+    double c_lon = std::cos(pos_lla.longitude * DEG_TO_RAD);
+    double s_lon = std::sin(pos_lla.longitude * DEG_TO_RAD);
+
+    double x_ecef = (r_curve + pos_lla.altitude) * c_lat * c_lon;
+    double y_ecef = (r_curve + pos_lla.altitude) * c_lat * s_lon;
+    double z_ecef = ((std::pow(b, 2) * r_curve / std::pow(a, 2))
+                        + pos_lla.altitude) * s_lat;
+
+    return PositionVelocityECEF {pos_lla.t, x_ecef, y_ecef, z_ecef};
+}
+
+/**
+ * get_rad_of_curvature
+ * @brief Calculates radius of curvature for conversion between
+ * LLA and ECEF position.
+ *
+ * Uses the formula:
+ *
+ *    r_curve = a / sqrt[1 - ( (e * e) * (sin(lat) * sin(lat)) )]
+ *
+ * For more information see Datum Transformation of GPS Positions, at:
+ * https://community.ptc.com/sejnu66972/attachments/sejnu66972/PTCMathcad/57177/2/Datum%20Transformations.pdf
+ *
+ * @param a         semi-major axis of the reference ellipsoid [meters]
+ * @param e         eccentricity of the reference ellipsoid [meters]
+ * @param latitude  [degrees]
+ *
+ * @return the calculated radius of curvature in meters
+ */
+double get_rad_of_curvature(double a, double e, double latitude) {
+    return (a / std::sqrt(1 - (std::pow(e, 2) *
+        std::pow(std::sin(latitude * DEG_TO_RAD), 2))));
+}
+
+/**
+ * get_vel_ecef
+ * @brief Calculates velocity at current point given previous point.
+ *
+ * Velocity is calculated as change in position over change in time
+ * between the given points.
+ *
+ * Populates the velocity fields of the current PositionVelocityECEF struct
+ * with the results of the calculation.
+ *
+ * @param previous  timestamped ECEF position at previous point
+ * @param current   timestamped ECEF position at which to calculate velocity
+ *
+ * @throws Runtime error if previous point is not before current point.
+ */
+void get_vel_ecef(
+    PositionVelocityECEF previous, PositionVelocityECEF* current) {
+    if (!ts_is_before(previous.t, current->t)) {
+      throw std::runtime_error(std::format(
```

Comment:
will this crash the program?

Reply:
could we just stop this execution and continue to the next?

Reply:
probably overkill for this project tbh

Reply:
I think I may have already done overkill so I'll leave it